content/articles/ansible-and-nix.md

```diff
@@ -0,0 +1,333 @@
+Distribution](http://nixos.org/~eelco/pubs/nixos-jfp-final.pdf) paper.
+
+For more practical questions, I found that the best way to learn was to find a
+package that resembles what I wanted to do and get inspiration from its
+corresponding expression. NixOS packages can be found using the [Search NixOS
+packages page](https://nixos.org/nixos/packages.html).
+
+
+## Comparing the deployment of a test application
+
+Using the [test
+project](https://github.com/WeAreWizards/voting-deployment-test), I am going to
+illustrate the practical differences between using NixOps and Ansible.
+
+### Dependencies
+
+Nix is first and foremost a package manager, so the most common way to deploy
+applications is to build packages locally and copy them to servers.
+
+Nix being `purely functional`, it should build the same output (a closure)
+given the same input (an expression). To do so, Nix needs all dependencies
+retrieved at build time.
+
+Assuming the dependencies you need are present in the nixos channel and at the
+right version then you can just refer to them like in the main derivation of
+the test project:
+
+```nix
+pythonPackages.buildPythonPackage {
+    name = "voting";
+
+    propagatedBuildInputs = [
+        pkgs.python27Packages.django
+        pkgs.python27Packages.sqlite3
+        pkgs.python27Packages.gunicorn
+    ];
+
+    src = ./.;
+}
+```
+
+However, if your dependencies are not in the nixos channel, you will need to
+declare them manually.
+
+For example the requests python library is declared like this:
+
+```nix
+requests = buildPythonPackage rec {
+  name = "requests-1.2.3";
+
+  src = pkgs.fetchurl {
+    url = "http://pypi.python.org/packages/source/r/requests/${name}.tar.gz";
+    md5 = "adbd3f18445f7fe5e77f65c502e264fb";
+  };
+
+  meta = {
+    description = "An Apache2 licensed HTTP library, written in Python, for human beings";
+    homepage = http://docs.python-requests.org/en/latest/;
+  };
+};
+```
+
+Which means than when building the package you are still dependent on pypi
+being up, unless the package is already cached.
+You can notice that the checksum of the archive is checked, so Nix will
+only build if the source package hasn't changed.
+
+Compared to a `requirements.txt` file, this is clearly more work. Fortunately
+the pypi2nix can take a `requirements.txt` file and write the corresponding
+derivations.
+
+Ansible doesn't have a single way of the handling of application dependencies.
+A common technique for a Python application is to install system packages with
+the distribution packages and python dependencies through pip.
+
+In the case of the voting test project this looks like this:
+
+```yaml
+    - name: global packages
+      apt: name={{item}} state=installed
+      with_items:
+        - python-virtualenv
+        - git
+        - uwsgi
+        - uwsgi-plugin-python
+        - nginx
+      sudo: yes
+
+    - name: python dependencies
+      pip: name=django version=1.6.6
```

Comment:
that's a good point. Django is at the version 1.6.6 in the nix channel so I used this version to keep things simple.

Earlier in the article, I point out the case "if your dependencies are not in the nixos channel".